### PR TITLE
Added application block for DefinitionBlocks and

### PR DESCRIPTION
**Pull request by Flumble**
modified expression generation to accommodate for them.

See DefinitionFunction, DefinitionBlock and FunctionBlock for the creation of an application.
The rest is a change from passing required Blocks to required OutputAnchors while generating an expression.

### CODE REVIEW

File: Code/src/main/java/nl/utwente/viskell/ui/components/FunctionBlock.java [hunk 1]
```diff
@@ -125,17 +126,21 @@ public final Integer getKnotIndex() {
...
         for (InputAnchor in : this.getActiveInputs()) {
-            Pair<Expression, Set<Block>> pair = in.getLocalExpr();
+            Pair<Expression, Set<OutputAnchor>> pair = in.getLocalExpr();
             expr = new Apply(expr, pair.a);
-            addLater.addAll(pair.b);
+            outsideAnchors.addAll(pair.b);
         }
         
-        return new Pair<>(expr, addLater);
+        if (funInfo instanceof DefinitionFunction) {
```
Comment (comonoidial) [inline Code/src/main/java/nl/utwente/viskell/ui/components/FunctionBlock.java]: Using instanceof here is a bit ugly, what about adding an extra method to FunctionInfo?

**Comment by comonoidial** [comment]
When building my own map function and applying it to an example it does work when I use the definition application.
But when I use the external output anchor of the definition block instead, I get ghci out of scope errors.
An example of the generated code:
```haskell
let {res__c150b8d = (((:) res__5ab3a4d4) res__1d069c48); lam__6f785057 = (\ map'_x0__2a9f64dd map'_x1__4920cb7e -> let {choiceoutput__2ff98b57 = case (()) of {() | (:) res0__e724b60 res1__3105d392 <- map'_x1__4920cb7e, res__1d069c48 <- ((lam__6f785057 map'_x0__2a9f64dd) res1__3105d392), res__5ab3a4d4 <- ((($) map'_x0__2a9f64dd) res0__e724b60), res__c150b8d <- (((:) res__5ab3a4d4) res__1d069c48), True -> res__c150b8d; () | True -> map'_x1__4920cb7e; }; } in choiceoutput__2ff98b57); val__7e8f969c = ([1..5]); res__2f8c09aa = succ; res__75b021c6 = ((($) lam__6f785057) res__2f8c09aa); res__bc64c73 = ((($) res__75b021c6) val__7e8f969c); } in res__bc64c73
```
Here are in the first let binding res__5ab3a4d4 and res__1d069c48 out of scope.

**Comment by comonoidial** [comment]
It looks like this:
mapexample

File: Code/src/main/java/nl/utwente/viskell/haskell/env/FunctionInfo.java [hunk 2]
```diff
@@ -16,20 +24,38 @@
     public FunctionInfo(String name, Type signature) {
         this.name = name;
         this.signature = signature;
+        this.requiredBlocks = Collections.EMPTY_LIST;
```
Comment (wandernauta) [inline Code/src/main/java/nl/utwente/viskell/haskell/env/FunctionInfo.java]: Constructors can call other constructors in Java, so this should also work:

    this(name, signature, ImmutableList.of())

File: Code/src/main/java/nl/utwente/viskell/haskell/env/FunctionInfo.java [hunk 3]
```diff
@@ -16,20 +24,38 @@
...
     public final String getName() {
         return this.name;
     }
 
-    /**
-     * @return The a fresh copy of type signature of this function.
-     */
+    /** @return The name of this function used for the front-end. */
+    public String getDisplayName() {
+        return this.name;
```
Comment (wandernauta) [inline Code/src/main/java/nl/utwente/viskell/haskell/env/FunctionInfo.java]: `return getName()` would perhaps be nicer.

File: Code/src/main/java/nl/utwente/viskell/ui/components/Block.java [hunk 4]
```diff
@@ -153,20 +153,20 @@ public void handleConnectionChanges(boolean finalPhase) {
         }
     }
     
-    /** @return A pair containing the expression this block represents and a set of required blocks. */
-    public abstract Pair<Expression,Set<Block>> getLocalExpr();
+    /** @return A pair containing the expression this block represents and a set of out-of-reach OutputAnchors. */
```
Comment (wandernauta) [inline Code/src/main/java/nl/utwente/viskell/ui/components/Block.java]: Maybe explain in the Javadoc why getLocalExpr needs to know about out-of-reach OutputAnchors.

**Comment by wandernauta** [comment]
I haven't tested it yet, but this looks neat! :+1:

**Comment by wandernauta** [comment]
Seems to work quite well for the most part. Two questions:

- It appears that changes inside the definition do not (correctly) propagate to usages of the definition outside/do not cause the usages to be re-evaluated: 2015-12-09-130433_726x508_scrot

- There might be a problem with zero-argument functions/'constants': 
2015-12-09-130442_698x491_scrot

**Comment by Flumble** [comment]
Only the display refreshing bug (https://github.com/wandernauta/viskell/pull/300#issuecomment-163203805) is left and occurs in multiple forms (both related and unrelated to this PR).

**Comment by wandernauta** [comment]
Merged, thanks a lot!

**Pull request merged**